pkg/controllers/build/container_controller.go

```diff
@@ -129,8 +164,74 @@ func (r *ContainerReconciler) resolveTargetImage(ctx context.Context, log logr.L
 	return image, nil
 }
 
+func (r *ContainerReconciler) resolveDigestReference(ctx context.Context, log logr.Logger, ref name.Reference, container *buildv1alpha1.Container) (string, error) {
+	keychain, err := r.constructKeychain(ctx, log, container)
+	if err != nil {
+		return "", err
+	}
+
+	auth, err := keychain.Resolve(ref.Context().Registry)
+	if err != nil {
+		log.Error(err, "unable to resolve auth for registry", "registry", ref.Context().RegistryStr())
+		return "", err
+	}
+
+	img, err := remote.Image(ref, remote.WithAuth(auth))
+	if err != nil {
+		log.Error(err, "failed to read image", "image", ref.String())
+		return "", err
+	}
+
+	digest, err := img.Digest()
+	if err != nil {
+		log.Error(err, "failed to get image digest", "image", ref.String())
+		return "", err
+	}
+
+	return fmt.Sprintf("%s@%s", ref.Context().Name(), digest), nil
+}
+
+func (r *ContainerReconciler) constructKeychain(ctx context.Context, log logr.Logger, container *buildv1alpha1.Container) (gauthn.Keychain, error) {
+	var serviceAccount corev1.ServiceAccount
+	if err := r.Get(ctx, types.NamespacedName{Namespace: container.Namespace, Name: riffBuildServiceAccount}, &serviceAccount); err != nil {
+		if apierrs.IsNotFound(err) {
+			log.Info("service account not found", "service-account", riffBuildServiceAccount)
+			return gauthn.DefaultKeychain, nil
```

Comment:
We dynamically create the `riff-build` service account when we see a credential, function or application. Since containers now need the service account as well, we should also setup the service account when a container is created.

That logic, exists here:
https://github.com/projectriff/system/blob/1377163d1d9832dad87c73e33a9284af1ec3542d/pkg/controllers/build/credentials_controller.go#L252-L254

Here we should fail and requeue the reconciliation because the service account should become available shortly.